{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "083bdcae",
   "metadata": {},
   "source": [
    "# Loai Gamal Mohamed                                     20180206\n",
    "# Ahmed Kadry Abd El-Shafy                            20180018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "45172aea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: wikipedia in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (1.4.0)\n",
      "Requirement already satisfied: beautifulsoup4 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from wikipedia) (4.10.0)\n",
      "Requirement already satisfied: requests<3.0.0,>=2.0.0 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from wikipedia) (2.27.1)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from requests<3.0.0,>=2.0.0->wikipedia) (2.0.12)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from requests<3.0.0,>=2.0.0->wikipedia) (2021.10.8)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from requests<3.0.0,>=2.0.0->wikipedia) (1.26.9)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from requests<3.0.0,>=2.0.0->wikipedia) (3.3)\n",
      "Requirement already satisfied: soupsieve>1.2 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from beautifulsoup4->wikipedia) (2.3.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install wikipedia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bde26cfe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: nltk in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (3.7)\n",
      "Requirement already satisfied: tqdm in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from nltk) (4.64.0)\n",
      "Requirement already satisfied: click in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from nltk) (8.1.2)\n",
      "Requirement already satisfied: regex>=2021.8.3 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from nltk) (2022.3.15)\n",
      "Requirement already satisfied: joblib in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from nltk) (1.1.0)\n",
      "Requirement already satisfied: importlib-metadata in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from click->nltk) (4.11.3)\n",
      "Requirement already satisfied: colorama in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from click->nltk) (0.4.4)\n",
      "Requirement already satisfied: typing-extensions>=3.6.4 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from importlib-metadata->click->nltk) (4.1.1)\n",
      "Requirement already satisfied: zipp>=0.5 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from importlib-metadata->click->nltk) (3.8.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install nltk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e176fb7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gensim in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (4.1.2)\n",
      "Requirement already satisfied: scipy>=0.18.1 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from gensim) (1.7.3)\n",
      "Requirement already satisfied: numpy>=1.17.0 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from gensim) (1.21.5)\n",
      "Requirement already satisfied: smart-open>=1.8.1 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from gensim) (5.2.1)\n",
      "Requirement already satisfied: Cython==0.29.23 in c:\\users\\loaig\\appdata\\local\\programs\\python\\python37\\lib\\site-packages (from gensim) (0.29.23)\n"
     ]
    }
   ],
   "source": [
    "!pip install gensim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1e05d31a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to\n",
      "[nltk_data]     C:\\Users\\loaig\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import nltk\n",
    "nltk.download('stopwords')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "152ca56d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any', 'are', 'aren', \"aren't\", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below', 'between', 'both', 'but', 'by', 'can', 'couldn', \"couldn't\", 'd', 'did', 'didn', \"didn't\", 'do', 'does', 'doesn', \"doesn't\", 'doing', 'don', \"don't\", 'down', 'during', 'each', 'few', 'for', 'from', 'further', 'had', 'hadn', \"hadn't\", 'has', 'hasn', \"hasn't\", 'have', 'haven', \"haven't\", 'having', 'he', 'her', 'here', 'hers', 'herself', 'him', 'himself', 'his', 'how', 'i', 'if', 'in', 'into', 'is', 'isn', \"isn't\", 'it', \"it's\", 'its', 'itself', 'just', 'll', 'm', 'ma', 'me', 'mightn', \"mightn't\", 'more', 'most', 'mustn', \"mustn't\", 'my', 'myself', 'needn', \"needn't\", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or', 'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's', 'same', 'shan', \"shan't\", 'she', \"she's\", 'should', \"should've\", 'shouldn', \"shouldn't\", 'so', 'some', 'such', 't', 'than', 'that', \"that'll\", 'the', 'their', 'theirs', 'them', 'themselves', 'then', 'there', 'these', 'they', 'this', 'those', 'through', 'to', 'too', 'under', 'until', 'up', 've', 'very', 'was', 'wasn', \"wasn't\", 'we', 'were', 'weren', \"weren't\", 'what', 'when', 'where', 'which', 'while', 'who', 'whom', 'why', 'will', 'with', 'won', \"won't\", 'wouldn', \"wouldn't\", 'y', 'you', \"you'd\", \"you'll\", \"you're\", \"you've\", 'your', 'yours', 'yourself', 'yourselves']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "179"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from nltk.corpus import stopwords\n",
    "\n",
    "stops = sorted(set(stopwords.words('english')))\n",
    "print(stops)\n",
    "len(stops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d76776d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import wikipedia\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cf3de992",
   "metadata": {},
   "outputs": [],
   "source": [
    "wikipedia.set_lang('en')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "19d73d7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def text_preprocessing(\n",
    "    text: str,\n",
    "    punctuations = r'''!()-[]{};:=+′–'\"\\,<>./?@#$%^&*_“~''',\n",
    "    stop_words=stops\n",
    "    )->list:\n",
    "    \n",
    "    for x in text.lower(): \n",
    "        if x in punctuations: \n",
    "            text = text.replace(x, \"\")\n",
    "\n",
    "    # Removing words that have numbers in them\n",
    "    text = re.sub(r'\\w*\\d\\w*', '', text)\n",
    "\n",
    "    # Removing digits\n",
    "    text = re.sub(r'[0-9]+', '', text)\n",
    "\n",
    "    # Cleaning the whitespaces\n",
    "    text = re.sub(r'\\s+', ' ', text).strip()\n",
    "\n",
    "    # Setting every word to lower\n",
    "    text = text.lower()\n",
    "\n",
    "    # Converting all our text to a list \n",
    "    text = text.split(' ')\n",
    "\n",
    "    # Droping empty strings\n",
    "    text = [x for x in text if x!='']\n",
    "\n",
    "    # Droping stop words\n",
    "    text = [x for x in text if x not in stop_words]\n",
    "\n",
    "    return text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "9d069495",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31530,\n",
       " 'Coronaviruses are a group of related RNA viruses that cause diseases in mammals and birds. In humans and birds, they cause respiratory tract infections that can range from mild to lethal. Mild illnesses in humans include some cases of the common cold (which is also caused by other viruses, predominantly rhinoviruses), while more lethal varieties can cause SARS, MERS and COVID-19, which is causing an ongoing pandemic. In cows and pigs they cause diarrhea, while in mice they cause hepatitis and encephalomyelitis.\\nCoronaviruses constitute the subfamily Orthocoronavirinae, in the family Coronaviridae, order Nidovirales and realm Riboviria. They are enveloped viruses with a positive-sense single-stranded RNA genome and a nucleocapsid of helical symmetry. The genome size of coronaviruses ranges from approximately 26 to 32 kilobases, one of the largest among RNA viruses. They have characteristic club-shaped spikes that project from their surface, which in electron micrographs create an image reminiscent of the solar corona, from which their name derives.\\n\\n\\n== Etymology ==\\nThe name \"coronavirus\" is derived from Latin corona, meaning \"crown\" or \"wreath\", itself a borrowing from Greek κορώνη korṓnē, \"garland, wreath\". The name was coined by June Almeida and David Tyrrell who first observed and studied human coronaviruses. The word was first used in print in 1968 by an informal group of virologists in the journal Nature to designate the new family of viruses. The name refers to the characteristic appearance of virions (the infective form of the virus) by electron microscopy, which have a fringe of large, bulbous surface projections creating an image reminiscent of the solar corona or halo. This morphology is created by the viral spike peplomers, which are proteins on the surface of the virus.The scientific name Coronavirus was accepted as a genus name by the International Committee for the Nomenclature of Viruses (later renamed International Committee on Taxonomy of Viruses) in 1971. As the number of new species increased, the genus was split into four genera, namely Alphacoronavirus, Betacoronavirus, Deltacoronavirus, and Gammacoronavirus in 2009. The common name coronavirus is used to refer to any member of the subfamily Orthocoronavirinae. As of 2020, 45 species are officially recognised.\\n\\n\\n== History ==\\n\\nThe earliest reports of a coronavirus infection in animals occurred in the late 1920s, when an acute respiratory infection of domesticated chickens emerged in North America. Arthur Schalk and M.C. Hawn in 1931 made the first detailed report which described a new respiratory infection of chickens in North Dakota. The infection of new-born chicks was characterized by gasping and listlessness with high mortality rates of 40–90%. Leland David Bushnell and Carl Alfred Brandly isolated the virus that caused the infection in 1933. The virus was then known as infectious bronchitis virus (IBV). Charles D. Hudson and Fred Robert Beaudette cultivated the virus for the first time in 1937. The specimen came to be known as the Beaudette strain. In the late 1940s, two more animal coronaviruses, JHM that causes brain disease (murine encephalitis) and mouse hepatitis virus (MHV) that causes hepatitis in mice were discovered. It was not realized at the time that these three different viruses were related.Human coronaviruses were discovered in the 1960s using two different methods in the United Kingdom and the United States. E.C. Kendall, Malcolm Bynoe, and David Tyrrell working at the Common Cold Unit of the British Medical Research Council collected a unique common cold virus designated B814 in 1961. The virus could not be cultivated using standard techniques which had successfully cultivated rhinoviruses, adenoviruses and other known common cold viruses. In 1965, Tyrrell and Bynoe successfully cultivated the novel virus by serially passing it through organ culture of human embryonic trachea. The new cultivating method was introduced to the lab by Bertil Hoorn. The isolated virus when intranasally inoculated into volunteers caused a cold and was inactivated by ether which indicated it had a lipid envelope. Dorothy Hamre and John Procknow at the University of Chicago isolated a novel cold from medical students in 1962. They isolated and grew the virus in kidney tissue culture, designating it 229E. The novel virus caused a cold in volunteers and, like B814, was inactivated by ether.\\n\\nScottish virologist June Almeida at St Thomas\\' Hospital in London, collaborating with Tyrrell, compared the structures of IBV, B814 and 229E in 1967. Using electron microscopy the three viruses were shown to be morphologically related by their general shape and distinctive club-like spikes. A research group at the National Institute of Health the same year was able to isolate another member of this new group of viruses using organ culture and named one of the samples OC43 (OC for organ culture). Like B814, 229E, and IBV, the novel cold virus OC43 had distinctive club-like spikes when observed with the electron microscope.The IBV-like novel cold viruses were soon shown to be also morphologically related to the mouse hepatitis virus. This new group of viruses were named coronaviruses after their distinctive morphological appearance. Human coronavirus 229E and human coronavirus OC43 continued to be studied in subsequent decades. The coronavirus strain B814 was lost. It is not known which present human coronavirus it was. Other human coronaviruses have since been identified, including SARS-CoV in 2003, HCoV NL63 in 2003, HCoV HKU1 in 2004, MERS-CoV in 2013, and SARS-CoV-2 in 2019. There have also been a large number of animal coronaviruses identified since the 1960s.\\n\\n\\n== Microbiology ==\\n\\n\\n=== Structure ===\\n\\nCoronaviruses are large, roughly spherical particles with unique surface projections. Their size is highly variable with average diameters of 80 to 120 nm. Extreme sizes are known from 50 to 200 nm in diameter. The total molecular mass is on average 40,000 kDa. They are enclosed in an envelope embedded with a number of protein molecules. The lipid bilayer envelope, membrane proteins, and nucleocapsid protect the virus when it is outside the host cell.The viral envelope is made up of a lipid bilayer in which the membrane (M), envelope (E) and spike (S) structural proteins are anchored. The molar ratio of E:S:M in the lipid bilayer is approximately 1:20:300.  The E and M protein are the structural proteins that combined with the lipid bilayer to shape the viral envelope and maintain its size. S proteins are needed for interaction with the host cells. But human coronavirus NL63 is peculiar in that its M protein has the binding site for the host cell, and not its S protein. The diameter of the envelope is 85 nm. The envelope of the virus in electron micrographs appears as a distinct pair of electron-dense shells (shells that are relatively opaque to the electron beam used to scan the virus particle).The M protein is the main structural protein of the envelope that provides the overall shape and is a type III membrane protein. It consists of 218 to 263 Amino acid residues and forms a layer 7.8 nm thick. It has three domains, a short N-terminal ectodomain, a triple-spanning transmembrane domain, and a C-terminal endodomain. The C-terminal domain forms a matrix-like lattice that adds to the extra-thickness of the envelope. Different species can have either N- or O-linked glycans in their protein amino-terminal domain. The M protein is crucial during the assembly, budding, envelope formation, and pathogenesis stages of the virus lifecycle.The E proteins are minor structural proteins and highly variable in different species. There are only about 20 copies of the E protein molecule in a coronavirus particle. They are 8.4 to 12 kDa in size and are composed of 76 to 109 amino acids. They are integral proteins (i.e. embedded in the lipid layer) and have two domains namely a transmembrane domain and an extramembrane C-terminal domain. They are almost fully α-helical, with a single α-helical transmembrane domain, and form pentameric (five-molecular) ion channels in the lipid bilayer. They are responsible for virion assembly, intracellular trafficking and morphogenesis (budding).\\n\\nThe spikes are the most distinguishing feature of coronaviruses and are responsible for the corona- or halo-like surface. On average a coronavirus particle has 74 surface spikes. Each spike is about 20 nm long and is composed of a trimer of the S protein. The S protein is in turn composed of an S1 and S2 subunit. The homotrimeric S protein is a class I fusion protein which mediates the receptor binding and membrane fusion between the virus and host cell. The S1 subunit forms the head of the spike and has the receptor-binding domain (RBD). The S2 subunit forms the stem which anchors the spike in the viral envelope and on protease activation enables fusion. The two subunits remain noncovalently linked as they are exposed on the viral surface until they attach to the host cell membrane. In a functionally active state, three S1 are attached to two S2 subunits. The subunit complex is split into individual subunits when the virus binds and fuses with the host cell under the action of proteases such as cathepsin family and transmembrane protease serine 2 (TMPRSS2) of the host cell.\\n\\nS1 proteins are the most critical components in terms of infection. They are also the most variable components as they are responsible for host cell specificity. They possess two major domains named N-terminal domain (S1-NTD) and C-terminal domain (S1-CTD), both of which serve as the receptor-binding domains. The NTDs recognize and bind sugars on the surface of the host cell. An exception is the MHV NTD that binds to a protein receptor carcinoembryonic antigen-related cell adhesion molecule 1 (CEACAM1). S1-CTDs are responsible for recognizing different protein receptors such as angiotensin-converting enzyme 2 (ACE2), aminopeptidase N (APN), and dipeptidyl peptidase 4 (DPP4).A subset of coronaviruses (specifically the members of betacoronavirus subgroup A) also has a shorter spike-like surface protein called hemagglutinin esterase (HE). The HE proteins occur as homodimers composed of about 400 amino acid residues and are 40 to 50 kDa in size. They appear as tiny surface projections of 5 to 7 nm long embedded in between the spikes. They help in the attachment to and detachment from the host cell.Inside the envelope, there is the nucleocapsid, which is formed from multiple copies of the nucleocapsid (N) protein, which are bound to the positive-sense single-stranded RNA genome in a continuous Beads-on-a-string type conformation. N protein is a phosphoprotein of 43 to 50 kDa in size, and is divided into three conserved domains. The majority of the protein is made up of domains 1 and 2, which are typically rich in arginines and lysines. Domain 3 has a short carboxy terminal end and has a net negative charge due to excess of acidic over basic amino acid residues.\\n\\n\\n=== Genome ===\\n\\nCoronaviruses contain a positive-sense, single-stranded RNA genome. The genome size for coronaviruses ranges from 26.4 to 31.7 kilobases. The genome size is one of the largest among RNA viruses. The genome has a 5′ methylated cap and a 3′ polyadenylated tail.The genome organization for a coronavirus is 5′-leader-UTR-replicase (ORF1ab)-spike (S)-envelope (E)-membrane (M)-nucleocapsid (N)-3′UTR-poly (A) tail. The open reading frames 1a and 1b, which occupy the first two-thirds of the genome, encode the replicase polyprotein (pp1ab). The replicase polyprotein self cleaves to form 16 nonstructural proteins (nsp1–nsp16).The later reading frames encode the four major structural proteins: spike, envelope, membrane, and nucleocapsid. Interspersed between these reading frames are the reading frames for the accessory proteins. The number of accessory proteins and their function is unique depending on the specific coronavirus.\\n\\n\\n=== Replication cycle ===\\n\\n\\n==== Cell entry ====\\n\\nInfection begins when the viral spike protein attaches to its complementary host cell receptor. After attachment, a protease of the host cell cleaves and activates the receptor-attached spike protein. Depending on the host cell protease available, cleavage and activation allows the virus to enter the host cell by endocytosis or direct fusion of the viral envelope with the host membrane.\\n\\n\\n==== Genome translation ====\\nOn entry into the host cell, the virus particle is uncoated, and its genome enters the cell cytoplasm. The coronavirus RNA genome has a 5′ methylated cap and a 3′ polyadenylated tail, which allows it to act like a messenger RNA and be directly translated by the host cell\\'s ribosomes. The host ribosomes translate the initial overlapping open reading frames ORF1a and ORF1b of the virus genome into two large overlapping polyproteins, pp1a and pp1ab.The larger polyprotein pp1ab is a result of a -1 ribosomal frameshift caused by a slippery sequence (UUUAAAC) and a downstream RNA pseudoknot at the end of open reading frame ORF1a. The ribosomal frameshift allows for the continuous translation of ORF1a followed by ORF1b.The polyproteins have their own proteases,  PLpro (nsp3) and 3CLpro (nsp5), which cleave the polyproteins at different specific sites. The cleavage of polyprotein pp1ab yields 16 nonstructural proteins (nsp1 to nsp16). Product proteins include various replication  proteins such as RNA-dependent RNA polymerase (nsp12), RNA helicase (nsp13), and exoribonuclease (nsp14).\\n\\n\\n==== Replicase-transcriptase ====\\n\\nA number of the nonstructural proteins coalesce to form a multi-protein replicase-transcriptase complex (RTC). The main replicase-transcriptase protein is the RNA-dependent RNA polymerase (RdRp). It is directly involved in the replication and transcription of RNA from an RNA strand. The other nonstructural proteins in the complex assist in the replication and transcription process. The exoribonuclease nonstructural protein, for instance, provides extra fidelity to replication by providing a proofreading function which the RNA-dependent RNA polymerase lacks.Replication – One of the main functions of the complex is to replicate the viral genome. RdRp directly mediates the synthesis of negative-sense genomic RNA from the positive-sense genomic RNA. This is followed by the replication of positive-sense genomic RNA from the negative-sense genomic RNA.\\n\\nTranscription – The other important function of the complex is to transcribe the viral genome. RdRp directly mediates the synthesis of negative-sense subgenomic RNA molecules from the positive-sense genomic RNA. This process is followed by the transcription of these negative-sense subgenomic RNA molecules to their corresponding positive-sense mRNAs. The subgenomic mRNAs form a \"nested set\" which have a common 5\\'-head and partially duplicate 3\\'-end.Recombination – The replicase-transcriptase complex is also capable of genetic recombination when at least two viral genomes are present in the same infected cell.  RNA recombination appears to be a major driving force in determining genetic variability within a coronavirus species, the capability of a coronavirus species to jump from one host to another and, infrequently, in determining the emergence of novel coronaviruses.  The exact mechanism of recombination in coronaviruses is unclear, but likely involves template switching during genome replication.\\n\\n\\n==== Assembly and release ====\\nThe replicated positive-sense genomic RNA becomes the genome of the progeny viruses. The mRNAs are gene transcripts of the last third of the virus genome after the initial overlapping reading frame. These mRNAs are translated by the host\\'s ribosomes into the structural proteins and many accessory proteins. RNA translation occurs inside the endoplasmic reticulum. The viral structural proteins S, E, and M move along the secretory pathway into the Golgi intermediate compartment. There, the M proteins direct most protein-protein interactions required for the assembly of viruses following its binding to the nucleocapsid. Progeny viruses are then released from the host cell by exocytosis through secretory vesicles. Once released the viruses can infect other host cells.\\n\\n\\n=== Transmission ===\\nInfected carriers are able to shed viruses into the environment. The interaction of the coronavirus spike protein with its complementary cell receptor is central in determining the tissue tropism, infectivity, and species range of the released virus. Coronaviruses mainly target epithelial cells. They are transmitted from one host to another host, depending on the coronavirus species, by either an aerosol, fomite, or fecal-oral route.Human coronaviruses infect the epithelial cells of the respiratory tract, while animal coronaviruses generally infect the epithelial cells of the digestive tract. SARS coronavirus, for example, infects the human epithelial cells of the lungs via an aerosol route by binding to the angiotensin-converting enzyme 2 (ACE2) receptor. Transmissible gastroenteritis coronavirus (TGEV) infects the pig epithelial cells of the digestive tract via a fecal-oral route by binding to the Alanine aminopeptidase (APN) receptor.\\n\\n\\n== Classification ==\\n\\nCoronaviruses form the subfamily Orthocoronavirinae, which is one of two sub-families in the family Coronaviridae, order Nidovirales, and realm Riboviria. They are divided into the four genera: Alphacoronavirus, Betacoronavirus, Gammacoronavirus and Deltacoronavirus. Alphacoronaviruses and betacoronaviruses infect mammals, while gammacoronaviruses and deltacoronaviruses primarily infect birds.\\n\\nGenus: Alphacoronavirus;Species: Alphacoronavirus 1 (TGEV, Feline coronavirus, Canine coronavirus), Human coronavirus 229E, Human coronavirus NL63, Miniopterus bat coronavirus 1, Miniopterus bat coronavirus HKU8, Porcine epidemic diarrhea virus, Rhinolophus bat coronavirus HKU2, Scotophilus bat coronavirus 512\\nGenus Betacoronavirus;Species: Betacoronavirus 1 (Bovine Coronavirus, Human coronavirus OC43), Hedgehog coronavirus 1, Human coronavirus HKU1, Middle East respiratory syndrome-related coronavirus, Murine coronavirus, Pipistrellus bat coronavirus HKU5, Rousettus bat coronavirus HKU9, Severe acute respiratory syndrome–related coronavirus (SARS-CoV, SARS-CoV-2), Tylonycteris bat coronavirus HKU4\\nGenus Gammacoronavirus;Species: Avian coronavirus, Beluga whale coronavirus SW1\\nGenus Deltacoronavirus\\nSpecies: Bulbul coronavirus HKU11, Porcine coronavirus HKU15\\n\\n\\n== Origin ==\\n\\nThe most recent common ancestor (MRCA) of all coronaviruses is estimated to have existed as recently as 8000 BCE, although some models place the common ancestor as far back as 55 million years or more, implying long term coevolution with bat and avian species. The most recent common ancestor of the alphacoronavirus line has been placed at about 2400 BCE, of the betacoronavirus line at 3300 BCE, of the gammacoronavirus line at 2800 BCE, and the deltacoronavirus line at about 3000 BCE. Bats and birds, as warm-blooded flying vertebrates, are an ideal natural reservoir for the coronavirus gene pool (with bats the reservoir for alphacoronaviruses and betacoronavirus – and birds the reservoir for gammacoronaviruses and deltacoronaviruses). The large number and global range of bat and avian species that host viruses have enabled extensive evolution and dissemination of coronaviruses.Many human coronaviruses have their origin in bats. The human coronavirus NL63 shared a common ancestor with a bat coronavirus (ARCoV.2) between 1190 and 1449 CE. The human coronavirus 229E shared a common ancestor with a bat coronavirus (GhanaGrp1 Bt CoV) between 1686 and 1800 CE. More recently, alpaca coronavirus and human coronavirus 229E diverged sometime before 1960. MERS-CoV emerged in humans from bats through the intermediate host of camels. MERS-CoV, although related to several bat coronavirus species, appears to have diverged from these several centuries ago. The most closely related bat coronavirus and SARS-CoV diverged in 1986. The ancestors of SARS-CoV first infected leaf-nose bats of the genus Hipposideridae; subsequently, they spread to horseshoe bats in the species Rhinolophidae, then to Asian palm civets, and finally to humans.Unlike other betacoronaviruses, bovine coronavirus of the species Betacoronavirus 1 and subgenus Embecovirus is thought to have originated in rodents and not in bats. In the 1790s, equine coronavirus diverged from the bovine coronavirus after a cross-species jump. Later in the 1890s, human coronavirus OC43 diverged from bovine coronavirus after another cross-species spillover event. It is speculated that the flu pandemic of 1890 may have been caused by this spillover event, and not by the influenza virus, because of the related timing, neurological symptoms, and unknown causative agent of the pandemic. Besides causing respiratory infections, human coronavirus OC43 is also suspected of playing a role in neurological diseases. In the 1950s, the human coronavirus OC43 began to diverge into its present genotypes. Phylogenetically, mouse hepatitis virus (Murine coronavirus), which infects the mouse\\'s liver and central nervous system, is related to human coronavirus OC43 and bovine coronavirus. Human coronavirus HKU1, like the aforementioned viruses, also has its origins in rodents.\\n\\n\\n== Infection in humans ==\\n\\nCoronaviruses vary significantly in risk factor. Some can kill more than 30% of those infected, such as MERS-CoV, and some are relatively harmless, such as the common cold. Coronaviruses can cause colds with major symptoms, such as fever, and a sore throat from swollen adenoids. Coronaviruses can cause pneumonia (either direct viral pneumonia or secondary bacterial pneumonia) and bronchitis (either direct viral bronchitis or secondary bacterial bronchitis). The human coronavirus discovered in 2003, SARS-CoV, which causes severe acute respiratory syndrome (SARS), has a unique pathogenesis because it causes both upper and lower respiratory tract infections.Six species of human coronaviruses are known, with one species subdivided into two different strains, making seven strains of human coronaviruses altogether.\\n\\nFour human coronaviruses produce symptoms that are generally mild, even though it is contended they might have been more aggressive in the past:\\nHuman coronavirus OC43 (HCoV-OC43), β-CoV\\nHuman coronavirus HKU1 (HCoV-HKU1), β-CoV\\nHuman coronavirus 229E (HCoV-229E), α-CoV\\nHuman coronavirus NL63 (HCoV-NL63), α-CoV–Three human coronaviruses produce potentially severe symptoms:\\n\\nSevere acute respiratory syndrome coronavirus (SARS-CoV), β-CoV (identified in 2003)\\nMiddle East respiratory syndrome-related coronavirus (MERS-CoV), β-CoV (identified in 2012)\\nSevere acute respiratory syndrome coronavirus 2 (SARS-CoV-2), β-CoV (identified in 2019)These cause the diseases commonly called SARS, MERS, and COVID-19 respectively.\\n\\n\\n=== Common cold ===\\n\\nAlthough the common cold is usually caused by rhinoviruses, in about 15% of cases the cause is a coronavirus. The human coronaviruses HCoV-OC43, HCoV-HKU1, HCoV-229E, and HCoV-NL63 continually circulate in the human population in adults and children worldwide and produce the generally mild symptoms of the common cold. The four mild coronaviruses have a seasonal incidence occurring in the winter months in temperate climates.  There is no preponderance in any season in tropical climates.\\n\\n\\n=== Severe acute respiratory syndrome (SARS) ===\\n\\nIn 2003, following the outbreak of severe acute respiratory syndrome (SARS) which had begun the prior year in Asia, and secondary cases elsewhere in the world, the World Health Organization (WHO) issued a press release stating that a novel coronavirus identified by several laboratories was the causative agent for SARS. The virus was officially named the SARS coronavirus (SARS-CoV). More than 8,000 people from 29 countries and territories were infected, and at least 774 died.\\n\\n\\n=== Middle East respiratory syndrome (MERS) ===\\n\\nIn September 2012, a new type of coronavirus was identified, initially called Novel Coronavirus 2012, and now officially named Middle East respiratory syndrome coronavirus (MERS-CoV). The World Health Organization issued a global alert soon after. The WHO update on 28 September 2012 said the virus did not seem to pass easily from person to person. However, on 12 May 2013, a case of human-to-human transmission in France was confirmed by the French Ministry of Social Affairs and Health. In addition, cases of human-to-human transmission were reported by the Ministry of Health in Tunisia. Two confirmed cases involved people who seemed to have caught the disease from their late father, who became ill after a visit to Qatar and Saudi Arabia. Despite this, it appears the virus had trouble spreading from human to human, as most individuals who are infected do not transmit the virus. By 30 October 2013, there were 124 cases and 52 deaths in Saudi Arabia.After the Dutch Erasmus Medical Centre sequenced the virus, the virus was given a new name, Human Coronavirus–Erasmus Medical Centre (HCoV-EMC). The final name for the virus is Middle East respiratory syndrome coronavirus (MERS-CoV). The only U.S. cases (both survived) were recorded in May 2014.In May 2015, an outbreak of MERS-CoV occurred in the Republic of Korea, when a man who had traveled to the Middle East, visited four hospitals in the Seoul area to treat his illness. This caused one of the largest outbreaks of MERS-CoV outside the Middle East. As of December 2019, 2,468 cases of MERS-CoV infection had been confirmed by laboratory tests, 851 of which were fatal, a mortality rate of approximately 34.5%.\\n\\n\\n=== Coronavirus disease 2019 (COVID-19) ===\\n\\nIn December 2019, a pneumonia outbreak was reported in Wuhan, China. On 31 December 2019, the outbreak was traced to a novel strain of coronavirus, which was given the interim name 2019-nCoV by the World Health Organization, later renamed SARS-CoV-2 by the International Committee on Taxonomy of Viruses.\\nAs of 20 April 2022, there have been at least 6,206,609 confirmed deaths and more than 506,627,123 confirmed cases in the COVID-19 pandemic. The Wuhan strain has been identified as a new strain of Betacoronavirus from group 2B with approximately 70% genetic similarity to the SARS-CoV. The virus has a 96% similarity to a bat coronavirus, so it is widely suspected to originate from bats as well.\\n\\n\\n=== Coronavirus HuPn-2018 ===\\n\\nDuring a surveillance study of archived samples of Malaysian viral pneumonia patients, virologists identified a strain of canine coronavirus which has infected humans in 2018.\\n\\n\\n== Infection in animals ==\\nCoronaviruses have been recognized as causing pathological conditions in veterinary medicine since the 1930s.  They infect a range of animals including swine, cattle, horses, camels, cats, dogs, rodents, birds and bats.  The majority of animal related coronaviruses infect the intestinal tract and are transmitted by a fecal-oral route. Significant research efforts have been focused on elucidating the viral pathogenesis of these animal coronaviruses, especially by virologists interested in veterinary and zoonotic diseases.\\n\\n\\n=== Farm animals ===\\nCoronaviruses infect domesticated birds. Infectious bronchitis virus (IBV), a type of coronavirus, causes avian infectious bronchitis. The virus is of concern to the poultry industry because of the high mortality from infection, its rapid spread, and its effect on production. The virus affects both meat production and egg production and causes substantial economic loss. In chickens, infectious bronchitis virus targets not only the respiratory tract but also the urogenital tract. The virus can spread to different organs throughout the chicken. The virus is transmitted by aerosol and food contaminated by feces. Different vaccines against IBV exist and have helped to limit the spread of the virus and its variants. Infectious bronchitis virus is one of a number of strains of the species Avian coronavirus. Another strain of avian coronavirus is turkey coronavirus (TCV) which causes enteritis in turkeys.Coronaviruses also affect other branches of animal husbandry such as pig farming and the Cattle raising. Swine acute diarrhea syndrome coronavirus (SADS-CoV), which is related to bat coronavirus HKU2, causes diarrhea in pigs. Porcine epidemic diarrhea virus (PEDV) is a coronavirus that has recently emerged and similarly causes diarrhea in pigs. Transmissible gastroenteritis virus (TGEV), which is a member of the species Alphacoronavirus 1, is another coronavirus that causes diarrhea in young pigs. In the cattle industry bovine coronavirus (BCV), which is a member of the species Betacoronavirus 1 and related to HCoV-OC43, is responsible for severe profuse enteritis in young calves.\\n\\n\\n=== Domestic pets ===\\nCoronaviruses infect domestic pets such as cats, dogs, and ferrets. There are two forms of feline coronavirus which are both members of the species Alphacoronavirus 1. Feline enteric coronavirus is a pathogen of minor clinical significance, but spontaneous mutation of this virus can result in feline infectious peritonitis (FIP), a disease with high mortality. There are two different coronaviruses that infect dogs. Canine coronavirus (CCoV), which is a member of the species Alphacoronavirus 1, causes mild gastrointestinal disease. Canine respiratory coronavirus (CRCoV), which is a member of the species Betacoronavirus 1 and related to HCoV-OC43, cause respiratory disease. Similarly, there are two types of coronavirus that infect ferrets. Ferret enteric coronavirus causes a gastrointestinal syndrome known as epizootic catarrhal enteritis (ECE), and a more lethal systemic version of the virus (like FIP in cats) known as ferret systemic coronavirus (FSC).\\n\\n\\n=== Laboratory animals ===\\nCoronaviruses infect laboratory animals. Mouse hepatitis virus (MHV), which is a member of the species Murine coronavirus, causes an epidemic murine illness with high mortality, especially among colonies of laboratory mice.  Prior to the discovery of SARS-CoV, MHV was the best-studied coronavirus both in vivo and in vitro as well as at the molecular level.  Some strains of MHV cause a progressive demyelinating encephalitis in mice which has been used as a murine model for multiple sclerosis. Sialodacryoadenitis virus (SDAV), which is a strain of the species Murine coronavirus, is highly infectious coronavirus of laboratory rats, which can be transmitted between individuals by direct contact and indirectly by aerosol.  Rabbit enteric coronavirus causes acute gastrointestinal disease and diarrhea in young European rabbits.  Mortality rates are high.\\n\\n\\n== Prevention and treatment ==\\nA number of vaccines using different methods have been developed against human coronavirus SARS-CoV-2. Antiviral targets against human coronaviruses have also been identified such as viral proteases, polymerases, and entry proteins. Drugs are in development which target these proteins and the different steps of viral replication.Vaccines are available for animal coronaviruses IBV, TGEV, and Canine CoV, although their effectiveness is limited. In the case of outbreaks of highly contagious animal coronaviruses, such as PEDV, measures such as destruction of entire herds of pigs may be used to prevent transmission to other herds.\\n\\n\\n== See also ==\\nCoronavirus diseases\\nZoonosis\\n\\n\\n== References ==\\n\\n\\n== Further reading ==')"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covid = wikipedia.page(\"Coronavirus\")\n",
    "covid = covid.content\n",
    "len(covid), covid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "faae7dd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2692,\n",
       " ['coronaviruses',\n",
       "  'group',\n",
       "  'related',\n",
       "  'rna',\n",
       "  'viruses',\n",
       "  'cause',\n",
       "  'diseases',\n",
       "  'mammals',\n",
       "  'birds',\n",
       "  'humans',\n",
       "  'birds',\n",
       "  'cause',\n",
       "  'respiratory',\n",
       "  'tract',\n",
       "  'infections',\n",
       "  'range',\n",
       "  'mild',\n",
       "  'lethal',\n",
       "  'mild',\n",
       "  'illnesses',\n",
       "  'humans',\n",
       "  'include',\n",
       "  'cases',\n",
       "  'common',\n",
       "  'cold',\n",
       "  'also',\n",
       "  'caused',\n",
       "  'viruses',\n",
       "  'predominantly',\n",
       "  'rhinoviruses',\n",
       "  'lethal',\n",
       "  'varieties',\n",
       "  'cause',\n",
       "  'sars',\n",
       "  'mers',\n",
       "  'causing',\n",
       "  'ongoing',\n",
       "  'pandemic',\n",
       "  'cows',\n",
       "  'pigs',\n",
       "  'cause',\n",
       "  'diarrhea',\n",
       "  'mice',\n",
       "  'cause',\n",
       "  'hepatitis',\n",
       "  'encephalomyelitis',\n",
       "  'coronaviruses',\n",
       "  'constitute',\n",
       "  'subfamily',\n",
       "  'orthocoronavirinae',\n",
       "  'family',\n",
       "  'coronaviridae',\n",
       "  'order',\n",
       "  'nidovirales',\n",
       "  'realm',\n",
       "  'riboviria',\n",
       "  'enveloped',\n",
       "  'viruses',\n",
       "  'positivesense',\n",
       "  'singlestranded',\n",
       "  'rna',\n",
       "  'genome',\n",
       "  'nucleocapsid',\n",
       "  'helical',\n",
       "  'symmetry',\n",
       "  'genome',\n",
       "  'size',\n",
       "  'coronaviruses',\n",
       "  'ranges',\n",
       "  'approximately',\n",
       "  'kilobases',\n",
       "  'one',\n",
       "  'largest',\n",
       "  'among',\n",
       "  'rna',\n",
       "  'viruses',\n",
       "  'characteristic',\n",
       "  'clubshaped',\n",
       "  'spikes',\n",
       "  'project',\n",
       "  'surface',\n",
       "  'electron',\n",
       "  'micrographs',\n",
       "  'create',\n",
       "  'image',\n",
       "  'reminiscent',\n",
       "  'solar',\n",
       "  'corona',\n",
       "  'name',\n",
       "  'derives',\n",
       "  'etymology',\n",
       "  'name',\n",
       "  'coronavirus',\n",
       "  'derived',\n",
       "  'latin',\n",
       "  'corona',\n",
       "  'meaning',\n",
       "  'crown',\n",
       "  'wreath',\n",
       "  'borrowing',\n",
       "  'greek',\n",
       "  'κορώνη',\n",
       "  'korṓnē',\n",
       "  'garland',\n",
       "  'wreath',\n",
       "  'name',\n",
       "  'coined',\n",
       "  'june',\n",
       "  'almeida',\n",
       "  'david',\n",
       "  'tyrrell',\n",
       "  'first',\n",
       "  'observed',\n",
       "  'studied',\n",
       "  'human',\n",
       "  'coronaviruses',\n",
       "  'word',\n",
       "  'first',\n",
       "  'used',\n",
       "  'print',\n",
       "  'informal',\n",
       "  'group',\n",
       "  'virologists',\n",
       "  'journal',\n",
       "  'nature',\n",
       "  'designate',\n",
       "  'new',\n",
       "  'family',\n",
       "  'viruses',\n",
       "  'name',\n",
       "  'refers',\n",
       "  'characteristic',\n",
       "  'appearance',\n",
       "  'virions',\n",
       "  'infective',\n",
       "  'form',\n",
       "  'virus',\n",
       "  'electron',\n",
       "  'microscopy',\n",
       "  'fringe',\n",
       "  'large',\n",
       "  'bulbous',\n",
       "  'surface',\n",
       "  'projections',\n",
       "  'creating',\n",
       "  'image',\n",
       "  'reminiscent',\n",
       "  'solar',\n",
       "  'corona',\n",
       "  'halo',\n",
       "  'morphology',\n",
       "  'created',\n",
       "  'viral',\n",
       "  'spike',\n",
       "  'peplomers',\n",
       "  'proteins',\n",
       "  'surface',\n",
       "  'virusthe',\n",
       "  'scientific',\n",
       "  'name',\n",
       "  'coronavirus',\n",
       "  'accepted',\n",
       "  'genus',\n",
       "  'name',\n",
       "  'international',\n",
       "  'committee',\n",
       "  'nomenclature',\n",
       "  'viruses',\n",
       "  'later',\n",
       "  'renamed',\n",
       "  'international',\n",
       "  'committee',\n",
       "  'taxonomy',\n",
       "  'viruses',\n",
       "  'number',\n",
       "  'new',\n",
       "  'species',\n",
       "  'increased',\n",
       "  'genus',\n",
       "  'split',\n",
       "  'four',\n",
       "  'genera',\n",
       "  'namely',\n",
       "  'alphacoronavirus',\n",
       "  'betacoronavirus',\n",
       "  'deltacoronavirus',\n",
       "  'gammacoronavirus',\n",
       "  'common',\n",
       "  'name',\n",
       "  'coronavirus',\n",
       "  'used',\n",
       "  'refer',\n",
       "  'member',\n",
       "  'subfamily',\n",
       "  'orthocoronavirinae',\n",
       "  'species',\n",
       "  'officially',\n",
       "  'recognised',\n",
       "  'history',\n",
       "  'earliest',\n",
       "  'reports',\n",
       "  'coronavirus',\n",
       "  'infection',\n",
       "  'animals',\n",
       "  'occurred',\n",
       "  'late',\n",
       "  'acute',\n",
       "  'respiratory',\n",
       "  'infection',\n",
       "  'domesticated',\n",
       "  'chickens',\n",
       "  'emerged',\n",
       "  'north',\n",
       "  'america',\n",
       "  'arthur',\n",
       "  'schalk',\n",
       "  'mc',\n",
       "  'hawn',\n",
       "  'made',\n",
       "  'first',\n",
       "  'detailed',\n",
       "  'report',\n",
       "  'described',\n",
       "  'new',\n",
       "  'respiratory',\n",
       "  'infection',\n",
       "  'chickens',\n",
       "  'north',\n",
       "  'dakota',\n",
       "  'infection',\n",
       "  'newborn',\n",
       "  'chicks',\n",
       "  'characterized',\n",
       "  'gasping',\n",
       "  'listlessness',\n",
       "  'high',\n",
       "  'mortality',\n",
       "  'rates',\n",
       "  'leland',\n",
       "  'david',\n",
       "  'bushnell',\n",
       "  'carl',\n",
       "  'alfred',\n",
       "  'brandly',\n",
       "  'isolated',\n",
       "  'virus',\n",
       "  'caused',\n",
       "  'infection',\n",
       "  'virus',\n",
       "  'known',\n",
       "  'infectious',\n",
       "  'bronchitis',\n",
       "  'virus',\n",
       "  'ibv',\n",
       "  'charles',\n",
       "  'hudson',\n",
       "  'fred',\n",
       "  'robert',\n",
       "  'beaudette',\n",
       "  'cultivated',\n",
       "  'virus',\n",
       "  'first',\n",
       "  'time',\n",
       "  'specimen',\n",
       "  'came',\n",
       "  'known',\n",
       "  'beaudette',\n",
       "  'strain',\n",
       "  'late',\n",
       "  'two',\n",
       "  'animal',\n",
       "  'coronaviruses',\n",
       "  'jhm',\n",
       "  'causes',\n",
       "  'brain',\n",
       "  'disease',\n",
       "  'murine',\n",
       "  'encephalitis',\n",
       "  'mouse',\n",
       "  'hepatitis',\n",
       "  'virus',\n",
       "  'mhv',\n",
       "  'causes',\n",
       "  'hepatitis',\n",
       "  'mice',\n",
       "  'discovered',\n",
       "  'realized',\n",
       "  'time',\n",
       "  'three',\n",
       "  'different',\n",
       "  'viruses',\n",
       "  'relatedhuman',\n",
       "  'coronaviruses',\n",
       "  'discovered',\n",
       "  'using',\n",
       "  'two',\n",
       "  'different',\n",
       "  'methods',\n",
       "  'united',\n",
       "  'kingdom',\n",
       "  'united',\n",
       "  'states',\n",
       "  'ec',\n",
       "  'kendall',\n",
       "  'malcolm',\n",
       "  'bynoe',\n",
       "  'david',\n",
       "  'tyrrell',\n",
       "  'working',\n",
       "  'common',\n",
       "  'cold',\n",
       "  'unit',\n",
       "  'british',\n",
       "  'medical',\n",
       "  'research',\n",
       "  'council',\n",
       "  'collected',\n",
       "  'unique',\n",
       "  'common',\n",
       "  'cold',\n",
       "  'virus',\n",
       "  'designated',\n",
       "  'virus',\n",
       "  'could',\n",
       "  'cultivated',\n",
       "  'using',\n",
       "  'standard',\n",
       "  'techniques',\n",
       "  'successfully',\n",
       "  'cultivated',\n",
       "  'rhinoviruses',\n",
       "  'adenoviruses',\n",
       "  'known',\n",
       "  'common',\n",
       "  'cold',\n",
       "  'viruses',\n",
       "  'tyrrell',\n",
       "  'bynoe',\n",
       "  'successfully',\n",
       "  'cultivated',\n",
       "  'novel',\n",
       "  'virus',\n",
       "  'serially',\n",
       "  'passing',\n",
       "  'organ',\n",
       "  'culture',\n",
       "  'human',\n",
       "  'embryonic',\n",
       "  'trachea',\n",
       "  'new',\n",
       "  'cultivating',\n",
       "  'method',\n",
       "  'introduced',\n",
       "  'lab',\n",
       "  'bertil',\n",
       "  'hoorn',\n",
       "  'isolated',\n",
       "  'virus',\n",
       "  'intranasally',\n",
       "  'inoculated',\n",
       "  'volunteers',\n",
       "  'caused',\n",
       "  'cold',\n",
       "  'inactivated',\n",
       "  'ether',\n",
       "  'indicated',\n",
       "  'lipid',\n",
       "  'envelope',\n",
       "  'dorothy',\n",
       "  'hamre',\n",
       "  'john',\n",
       "  'procknow',\n",
       "  'university',\n",
       "  'chicago',\n",
       "  'isolated',\n",
       "  'novel',\n",
       "  'cold',\n",
       "  'medical',\n",
       "  'students',\n",
       "  'isolated',\n",
       "  'grew',\n",
       "  'virus',\n",
       "  'kidney',\n",
       "  'tissue',\n",
       "  'culture',\n",
       "  'designating',\n",
       "  'novel',\n",
       "  'virus',\n",
       "  'caused',\n",
       "  'cold',\n",
       "  'volunteers',\n",
       "  'like',\n",
       "  'inactivated',\n",
       "  'ether',\n",
       "  'scottish',\n",
       "  'virologist',\n",
       "  'june',\n",
       "  'almeida',\n",
       "  'st',\n",
       "  'thomas',\n",
       "  'hospital',\n",
       "  'london',\n",
       "  'collaborating',\n",
       "  'tyrrell',\n",
       "  'compared',\n",
       "  'structures',\n",
       "  'ibv',\n",
       "  'using',\n",
       "  'electron',\n",
       "  'microscopy',\n",
       "  'three',\n",
       "  'viruses',\n",
       "  'shown',\n",
       "  'morphologically',\n",
       "  'related',\n",
       "  'general',\n",
       "  'shape',\n",
       "  'distinctive',\n",
       "  'clublike',\n",
       "  'spikes',\n",
       "  'research',\n",
       "  'group',\n",
       "  'national',\n",
       "  'institute',\n",
       "  'health',\n",
       "  'year',\n",
       "  'able',\n",
       "  'isolate',\n",
       "  'another',\n",
       "  'member',\n",
       "  'new',\n",
       "  'group',\n",
       "  'viruses',\n",
       "  'using',\n",
       "  'organ',\n",
       "  'culture',\n",
       "  'named',\n",
       "  'one',\n",
       "  'samples',\n",
       "  'oc',\n",
       "  'organ',\n",
       "  'culture',\n",
       "  'like',\n",
       "  'ibv',\n",
       "  'novel',\n",
       "  'cold',\n",
       "  'virus',\n",
       "  'distinctive',\n",
       "  'clublike',\n",
       "  'spikes',\n",
       "  'observed',\n",
       "  'electron',\n",
       "  'microscopethe',\n",
       "  'ibvlike',\n",
       "  'novel',\n",
       "  'cold',\n",
       "  'viruses',\n",
       "  'soon',\n",
       "  'shown',\n",
       "  'also',\n",
       "  'morphologically',\n",
       "  'related',\n",
       "  'mouse',\n",
       "  'hepatitis',\n",
       "  'virus',\n",
       "  'new',\n",
       "  'group',\n",
       "  'viruses',\n",
       "  'named',\n",
       "  'coronaviruses',\n",
       "  'distinctive',\n",
       "  'morphological',\n",
       "  'appearance',\n",
       "  'human',\n",
       "  'coronavirus',\n",
       "  'human',\n",
       "  'coronavirus',\n",
       "  'continued',\n",
       "  'studied',\n",
       "  'subsequent',\n",
       "  'decades',\n",
       "  'coronavirus',\n",
       "  'strain',\n",
       "  'lost',\n",
       "  'known',\n",
       "  'present',\n",
       "  'human',\n",
       "  'coronavirus',\n",
       "  'human',\n",
       "  'coronaviruses',\n",
       "  'since',\n",
       "  'identified',\n",
       "  'including',\n",
       "  'sarscov',\n",
       "  'hcov',\n",
       "  'hcov',\n",
       "  'merscov',\n",
       "  'also',\n",
       "  'large',\n",
       "  'number',\n",
       "  'animal',\n",
       "  'coronaviruses',\n",
       "  'identified',\n",
       "  'since',\n",
       "  'microbiology',\n",
       "  'structure',\n",
       "  'coronaviruses',\n",
       "  'large',\n",
       "  'roughly',\n",
       "  'spherical',\n",
       "  'particles',\n",
       "  'unique',\n",
       "  'surface',\n",
       "  'projections',\n",
       "  'size',\n",
       "  'highly',\n",
       "  'variable',\n",
       "  'average',\n",
       "  'diameters',\n",
       "  'nm',\n",
       "  'extreme',\n",
       "  'sizes',\n",
       "  'known',\n",
       "  'nm',\n",
       "  'diameter',\n",
       "  'total',\n",
       "  'molecular',\n",
       "  'mass',\n",
       "  'average',\n",
       "  'kda',\n",
       "  'enclosed',\n",
       "  'envelope',\n",
       "  'embedded',\n",
       "  'number',\n",
       "  'protein',\n",
       "  'molecules',\n",
       "  'lipid',\n",
       "  'bilayer',\n",
       "  'envelope',\n",
       "  'membrane',\n",
       "  'proteins',\n",
       "  'nucleocapsid',\n",
       "  'protect',\n",
       "  'virus',\n",
       "  'outside',\n",
       "  'host',\n",
       "  'cellthe',\n",
       "  'viral',\n",
       "  'envelope',\n",
       "  'made',\n",
       "  'lipid',\n",
       "  'bilayer',\n",
       "  'membrane',\n",
       "  'envelope',\n",
       "  'e',\n",
       "  'spike',\n",
       "  'structural',\n",
       "  'proteins',\n",
       "  'anchored',\n",
       "  'molar',\n",
       "  'ratio',\n",
       "  'esm',\n",
       "  'lipid',\n",
       "  'bilayer',\n",
       "  'approximately',\n",
       "  'e',\n",
       "  'protein',\n",
       "  'structural',\n",
       "  'proteins',\n",
       "  'combined',\n",
       "  'lipid',\n",
       "  'bilayer',\n",
       "  'shape',\n",
       "  'viral',\n",
       "  'envelope',\n",
       "  'maintain',\n",
       "  'size',\n",
       "  'proteins',\n",
       "  'needed',\n",
       "  'interaction',\n",
       "  'host',\n",
       "  'cells',\n",
       "  'human',\n",
       "  'coronavirus',\n",
       "  'peculiar',\n",
       "  'protein',\n",
       "  'binding',\n",
       "  'site',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'protein',\n",
       "  'diameter',\n",
       "  'envelope',\n",
       "  'nm',\n",
       "  'envelope',\n",
       "  'virus',\n",
       "  'electron',\n",
       "  'micrographs',\n",
       "  'appears',\n",
       "  'distinct',\n",
       "  'pair',\n",
       "  'electrondense',\n",
       "  'shells',\n",
       "  'shells',\n",
       "  'relatively',\n",
       "  'opaque',\n",
       "  'electron',\n",
       "  'beam',\n",
       "  'used',\n",
       "  'scan',\n",
       "  'virus',\n",
       "  'particlethe',\n",
       "  'protein',\n",
       "  'main',\n",
       "  'structural',\n",
       "  'protein',\n",
       "  'envelope',\n",
       "  'provides',\n",
       "  'overall',\n",
       "  'shape',\n",
       "  'type',\n",
       "  'iii',\n",
       "  'membrane',\n",
       "  'protein',\n",
       "  'consists',\n",
       "  'amino',\n",
       "  'acid',\n",
       "  'residues',\n",
       "  'forms',\n",
       "  'layer',\n",
       "  'nm',\n",
       "  'thick',\n",
       "  'three',\n",
       "  'domains',\n",
       "  'short',\n",
       "  'nterminal',\n",
       "  'ectodomain',\n",
       "  'triplespanning',\n",
       "  'transmembrane',\n",
       "  'domain',\n",
       "  'cterminal',\n",
       "  'endodomain',\n",
       "  'cterminal',\n",
       "  'domain',\n",
       "  'forms',\n",
       "  'matrixlike',\n",
       "  'lattice',\n",
       "  'adds',\n",
       "  'extrathickness',\n",
       "  'envelope',\n",
       "  'different',\n",
       "  'species',\n",
       "  'either',\n",
       "  'n',\n",
       "  'olinked',\n",
       "  'glycans',\n",
       "  'protein',\n",
       "  'aminoterminal',\n",
       "  'domain',\n",
       "  'protein',\n",
       "  'crucial',\n",
       "  'assembly',\n",
       "  'budding',\n",
       "  'envelope',\n",
       "  'formation',\n",
       "  'pathogenesis',\n",
       "  'stages',\n",
       "  'virus',\n",
       "  'lifecyclethe',\n",
       "  'e',\n",
       "  'proteins',\n",
       "  'minor',\n",
       "  'structural',\n",
       "  'proteins',\n",
       "  'highly',\n",
       "  'variable',\n",
       "  'different',\n",
       "  'species',\n",
       "  'copies',\n",
       "  'e',\n",
       "  'protein',\n",
       "  'molecule',\n",
       "  'coronavirus',\n",
       "  'particle',\n",
       "  'kda',\n",
       "  'size',\n",
       "  'composed',\n",
       "  'amino',\n",
       "  'acids',\n",
       "  'integral',\n",
       "  'proteins',\n",
       "  'ie',\n",
       "  'embedded',\n",
       "  'lipid',\n",
       "  'layer',\n",
       "  'two',\n",
       "  'domains',\n",
       "  'namely',\n",
       "  'transmembrane',\n",
       "  'domain',\n",
       "  'extramembrane',\n",
       "  'cterminal',\n",
       "  'domain',\n",
       "  'almost',\n",
       "  'fully',\n",
       "  'αhelical',\n",
       "  'single',\n",
       "  'αhelical',\n",
       "  'transmembrane',\n",
       "  'domain',\n",
       "  'form',\n",
       "  'pentameric',\n",
       "  'fivemolecular',\n",
       "  'ion',\n",
       "  'channels',\n",
       "  'lipid',\n",
       "  'bilayer',\n",
       "  'responsible',\n",
       "  'virion',\n",
       "  'assembly',\n",
       "  'intracellular',\n",
       "  'trafficking',\n",
       "  'morphogenesis',\n",
       "  'budding',\n",
       "  'spikes',\n",
       "  'distinguishing',\n",
       "  'feature',\n",
       "  'coronaviruses',\n",
       "  'responsible',\n",
       "  'corona',\n",
       "  'halolike',\n",
       "  'surface',\n",
       "  'average',\n",
       "  'coronavirus',\n",
       "  'particle',\n",
       "  'surface',\n",
       "  'spikes',\n",
       "  'spike',\n",
       "  'nm',\n",
       "  'long',\n",
       "  'composed',\n",
       "  'trimer',\n",
       "  'protein',\n",
       "  'protein',\n",
       "  'turn',\n",
       "  'composed',\n",
       "  'subunit',\n",
       "  'homotrimeric',\n",
       "  'protein',\n",
       "  'class',\n",
       "  'fusion',\n",
       "  'protein',\n",
       "  'mediates',\n",
       "  'receptor',\n",
       "  'binding',\n",
       "  'membrane',\n",
       "  'fusion',\n",
       "  'virus',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'subunit',\n",
       "  'forms',\n",
       "  'head',\n",
       "  'spike',\n",
       "  'receptorbinding',\n",
       "  'domain',\n",
       "  'rbd',\n",
       "  'subunit',\n",
       "  'forms',\n",
       "  'stem',\n",
       "  'anchors',\n",
       "  'spike',\n",
       "  'viral',\n",
       "  'envelope',\n",
       "  'protease',\n",
       "  'activation',\n",
       "  'enables',\n",
       "  'fusion',\n",
       "  'two',\n",
       "  'subunits',\n",
       "  'remain',\n",
       "  'noncovalently',\n",
       "  'linked',\n",
       "  'exposed',\n",
       "  'viral',\n",
       "  'surface',\n",
       "  'attach',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'membrane',\n",
       "  'functionally',\n",
       "  'active',\n",
       "  'state',\n",
       "  'three',\n",
       "  'attached',\n",
       "  'two',\n",
       "  'subunits',\n",
       "  'subunit',\n",
       "  'complex',\n",
       "  'split',\n",
       "  'individual',\n",
       "  'subunits',\n",
       "  'virus',\n",
       "  'binds',\n",
       "  'fuses',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'action',\n",
       "  'proteases',\n",
       "  'cathepsin',\n",
       "  'family',\n",
       "  'transmembrane',\n",
       "  'protease',\n",
       "  'serine',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'proteins',\n",
       "  'critical',\n",
       "  'components',\n",
       "  'terms',\n",
       "  'infection',\n",
       "  'also',\n",
       "  'variable',\n",
       "  'components',\n",
       "  'responsible',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'specificity',\n",
       "  'possess',\n",
       "  'two',\n",
       "  'major',\n",
       "  'domains',\n",
       "  'named',\n",
       "  'nterminal',\n",
       "  'domain',\n",
       "  'cterminal',\n",
       "  'domain',\n",
       "  'serve',\n",
       "  'receptorbinding',\n",
       "  'domains',\n",
       "  'ntds',\n",
       "  'recognize',\n",
       "  'bind',\n",
       "  'sugars',\n",
       "  'surface',\n",
       "  'host',\n",
       "  'cell',\n",
       "  'exception',\n",
       "  'mhv',\n",
       "  'ntd',\n",
       "  'binds',\n",
       "  'protein',\n",
       "  'receptor',\n",
       "  'carcinoembryonic',\n",
       "  'antigenrelated',\n",
       "  'cell',\n",
       "  'adhesion',\n",
       "  'molecule',\n",
       "  'responsible',\n",
       "  'recognizing',\n",
       "  'different',\n",
       "  'protein',\n",
       "  'receptors',\n",
       "  'angiotensinconverting',\n",
       "  'enzyme',\n",
       "  'aminopeptidase',\n",
       "  'n',\n",
       "  'apn',\n",
       "  'dipeptidyl',\n",
       "  'peptidase',\n",
       "  'subset',\n",
       "  'coronaviruses',\n",
       "  'specifically',\n",
       "  'members',\n",
       "  'betacoronavirus',\n",
       "  'subgroup',\n",
       "  'also',\n",
       "  'shorter',\n",
       "  'spikelike',\n",
       "  'surface',\n",
       "  'protein',\n",
       "  'called',\n",
       "  'hemagglutinin',\n",
       "  'esterase',\n",
       "  'proteins',\n",
       "  'occur',\n",
       "  'homodimers',\n",
       "  'composed',\n",
       "  'amino',\n",
       "  'acid',\n",
       "  'residues',\n",
       "  'kda',\n",
       "  'size',\n",
       "  'appear',\n",
       "  'tiny',\n",
       "  'surface',\n",
       "  'projections',\n",
       "  'nm',\n",
       "  'long',\n",
       "  'embedded',\n",
       "  'spikes',\n",
       "  'help',\n",
       "  'attachment',\n",
       "  'detachment',\n",
       "  'host',\n",
       "  'cellinside',\n",
       "  'envelope',\n",
       "  'nucleocapsid',\n",
       "  'formed',\n",
       "  'multiple',\n",
       "  'copies',\n",
       "  'nucleocapsid',\n",
       "  'n',\n",
       "  'protein',\n",
       "  'bound',\n",
       "  'positivesense',\n",
       "  'singlestranded',\n",
       "  'rna',\n",
       "  'genome',\n",
       "  'continuous',\n",
       "  'beadsonastring',\n",
       "  'type',\n",
       "  'conformation',\n",
       "  'n',\n",
       "  'protein',\n",
       "  'phosphoprotein',\n",
       "  'kda',\n",
       "  'size',\n",
       "  'divided',\n",
       "  'three',\n",
       "  'conserved',\n",
       "  'domains',\n",
       "  'majority',\n",
       "  'protein',\n",
       "  'made',\n",
       "  'domains',\n",
       "  'typically',\n",
       "  'rich',\n",
       "  'arginines',\n",
       "  'lysines',\n",
       "  'domain',\n",
       "  'short',\n",
       "  'carboxy',\n",
       "  'terminal',\n",
       "  'end',\n",
       "  'net',\n",
       "  'negative',\n",
       "  'charge',\n",
       "  'due',\n",
       "  'excess',\n",
       "  'acidic',\n",
       "  'basic',\n",
       "  'amino',\n",
       "  'acid',\n",
       "  'residues',\n",
       "  'genome',\n",
       "  'coronaviruses',\n",
       "  'contain',\n",
       "  'positivesense',\n",
       "  'singlestranded',\n",
       "  'rna',\n",
       "  'genome',\n",
       "  'genome',\n",
       "  'size',\n",
       "  'coronaviruses',\n",
       "  'ranges',\n",
       "  'kilobases',\n",
       "  'genome',\n",
       "  'size',\n",
       "  'one',\n",
       "  'largest',\n",
       "  'among',\n",
       "  'rna',\n",
       "  'viruses',\n",
       "  'genome',\n",
       "  'methylated',\n",
       "  'cap',\n",
       "  'polyadenylated',\n",
       "  'tailthe',\n",
       "  'genome',\n",
       "  'organization',\n",
       "  'coronavirus',\n",
       "  'senvelope',\n",
       "  'emembrane',\n",
       "  'mnucleocapsid',\n",
       "  'tail',\n",
       "  'open',\n",
       "  'reading',\n",
       "  'frames',\n",
       "  'occupy',\n",
       "  'first',\n",
       "  'twothirds',\n",
       "  'genome',\n",
       "  'encode',\n",
       "  'replicase',\n",
       "  'polyprotein',\n",
       "  'replicase',\n",
       "  'polyprotein',\n",
       "  'self',\n",
       "  'cleaves',\n",
       "  ...])"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covid = text_preprocessing(covid)\n",
    "len(covid), covid"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74a5099d",
   "metadata": {},
   "source": [
    "# Domains\n",
    "\n",
    "## 1- Operating systems (Windows, Linux, Android, MacOS)\n",
    "## 2- Football clubs (Al Ahly, Real Madrid, Liverpool, Juventus)\n",
    "## 3- Illnesses (The Flu, Allergy, Conjunctivitis, Headache)\n",
    "## 4- Cars (Mercedes-Benz, Toyota, Volvo, Nissan)\n",
    "## 5- Watch brands (Rolex SA, TAG Heuer S.A., Seiko Holdings Corporation, Compagnie des Montres Longines, Francillon S.A.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "3efeeb1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Windows\n",
      "Linux Kernel\n",
      "Android\n",
      "OSX\n",
      "Al Ahly\n",
      "Real Madrid\n",
      "Liverpool\n",
      "Juventus\n",
      "The Flu\n",
      "Allergy\n",
      "Conjunctivitis\n",
      "Headache\n",
      "Mercedes-Benz\n",
      "Toyota\n",
      "Volvo\n",
      "Nissan\n",
      "Rolex SA\n",
      "TAG Heuer S.A.\n",
      "Seiko Holdings Corporation\n",
      "Compagnie des Montres Longines, Francillon S.A.\n"
     ]
    }
   ],
   "source": [
    "docs = []\n",
    "titles = []\n",
    "for i in range(20):\n",
    "    val = input()\n",
    "    temp = wikipedia.page(val)\n",
    "    temp = temp.content\n",
    "    temp = text_preprocessing(temp)\n",
    "    docs.append(temp)\n",
    "    titles.append(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "971e17e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0f8b3a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gensim\n",
    "model = gensim.models.KeyedVectors.load_word2vec_format(\"./GoogleNews-vectors-negative300.bin\", binary=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f61dcd2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_embedding_vector(doc):\n",
    "    temp = []\n",
    "    for i in range(len(doc)):\n",
    "        if doc[i] in model:\n",
    "            vector = model[doc[i]]\n",
    "            temp.append(vector)\n",
    "        \n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "    temp = sum(temp) / len(temp)\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "f8c7c2c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "docs_embedding_vectors = []\n",
    "for doc in docs:\n",
    "    temp = get_embedding_vector(doc)\n",
    "    docs_embedding_vectors.append(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "57dc339d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20, (300,))"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(docs_embedding_vectors), docs_embedding_vectors[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "65dea11a",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Embedding_Vectors.txt', 'w') as f:\n",
    "    for vector in docs_embedding_vectors:\n",
    "        f.write(\"%s\\n\" % vector)\n",
    "        \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "73a2f57f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import distance\n",
    "\n",
    "def calculate_distances(test_embedding, embedding, titles):\n",
    "\n",
    "    similars = []\n",
    "    indexes = []\n",
    "    for idx, embedding in enumerate(docs_embedding_vectors):\n",
    "        dist = distance.euclidean(test_embedding, embedding)\n",
    "        similars.append(dist)\n",
    "        indexes.append(idx)\n",
    "        \n",
    "    for j in range(len(similars)):\n",
    "        for i in range(len(similars) - 1):\n",
    "            if similars[i] > similars[i + 1]:\n",
    "                similars[i],similars[i + 1] = similars[i + 1], similars[i]\n",
    "                indexes[i],indexes[i + 1] = indexes[i + 1], indexes[i]\n",
    "                \n",
    "    result = []\n",
    "    for i in range(len(indexes)):\n",
    "        j = indexes[i]\n",
    "        temp = titles[j]\n",
    "        result.append(temp)\n",
    "        \n",
    "    return similars, result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff0c5579",
   "metadata": {},
   "source": [
    "## Test with sentence about Covid19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "917d2f84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter Test Sentence: Coronaviruses are a group of related RNA viruses that cause diseases in mammals and birds. In humans and birds, they cause respiratory tract infections that can range from mild to lethal.\n"
     ]
    }
   ],
   "source": [
    "test = input(\"Enter Test Sentence: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "e21e31cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = text_preprocessing(test)\n",
    "test_embedding = get_embedding_vector(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e78944ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Flu \t 0.9300745725631714\n",
      "Conjunctivitis \t 1.0439094305038452\n",
      "Headache \t 1.2719266414642334\n",
      "Seiko Holdings Corporation \t 1.468782663345337\n",
      "Volvo \t 1.4705588817596436\n",
      "Rolex SA \t 1.473665475845337\n",
      "Liverpool \t 1.4767154455184937\n",
      "Mercedes-Benz \t 1.4860764741897583\n",
      "Android \t 1.4861050844192505\n",
      "Toyota \t 1.4904351234436035\n",
      "Compagnie des Montres Longines, Francillon S.A. \t 1.4905600547790527\n",
      "Linux Kernel \t 1.504093050956726\n",
      "TAG Heuer S.A. \t 1.5084986686706543\n",
      "OSX \t 1.5301265716552734\n",
      "Nissan \t 1.544398307800293\n",
      "Windows \t 1.5540870428085327\n",
      "Real Madrid \t 1.581506609916687\n",
      "Juventus \t 1.592857003211975\n",
      "Allergy \t 1.5962235927581787\n",
      "Al Ahly \t 1.6110955476760864\n"
     ]
    }
   ],
   "source": [
    "similars, result = calculate_distances(test_embedding, docs_embedding_vectors, titles)\n",
    "for i in range(len(similars)):\n",
    "    print(result[i], '\\t',  similars[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cdd032b",
   "metadata": {},
   "source": [
    "## Test with sentence about clubs (Zamalek)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "67a3bf99",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter Test Sentence: Zamalek Sporting Club, commonly referred to as Zamalek, is an Egyptian sports club based in Cairo. The club is mainly known for its professional football team, which plays in the Egyptian Premier League, the top tier of the Egyptian football league system.\n"
     ]
    }
   ],
   "source": [
    "test = input(\"Enter Test Sentence: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2eb64b24",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = text_preprocessing(test)\n",
    "test_embedding = get_embedding_vector(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "7a889edf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Juventus \t 0.6691205501556396\n",
      "Real Madrid \t 0.7171584963798523\n",
      "Al Ahly \t 0.7327566742897034\n",
      "Volvo \t 0.7722194194793701\n",
      "Liverpool \t 0.8456934690475464\n",
      "Compagnie des Montres Longines, Francillon S.A. \t 0.9007419347763062\n",
      "Seiko Holdings Corporation \t 0.9044776558876038\n",
      "TAG Heuer S.A. \t 0.9435185194015503\n",
      "Rolex SA \t 0.9534878134727478\n",
      "Mercedes-Benz \t 0.9715715050697327\n",
      "Nissan \t 1.0076216459274292\n",
      "Toyota \t 1.0134292840957642\n",
      "OSX \t 1.0850969552993774\n",
      "Android \t 1.0885872840881348\n",
      "Linux Kernel \t 1.1028143167495728\n",
      "Windows \t 1.1378031969070435\n",
      "Headache \t 1.2406426668167114\n",
      "Allergy \t 1.2482576370239258\n",
      "The Flu \t 1.2621780633926392\n",
      "Conjunctivitis \t 1.3251302242279053\n"
     ]
    }
   ],
   "source": [
    "similars, result = calculate_distances(test_embedding, docs_embedding_vectors, titles)\n",
    "for i in range(len(similars)):\n",
    "    print(result[i], '\\t',  similars[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9a7fe75",
   "metadata": {},
   "source": [
    "## Testing with sentence about cars (Bently)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4c629894",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter Test Sentence: Bentley Motors Limited is a British manufacturer and marketer of luxury cars and SUVs, and a subsidiary of the Volkswagen Group since 1998.[14]  Headquartered in Crewe, England, the company was founded as Bentley Motors Limited by W. O. Bentley in 1919 in Cricklewood, North London, and became widely known for winning the 24 Hours of Le Mans in 1924, 1927, 1928, 1929 and 1930.\n"
     ]
    }
   ],
   "source": [
    "test = input(\"Enter Test Sentence: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "b2650bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = text_preprocessing(test)\n",
    "test_embedding = get_embedding_vector(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b7d997e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nissan \t 0.6112778186798096\n",
      "Toyota \t 0.670693576335907\n",
      "Mercedes-Benz \t 0.6960440278053284\n",
      "Compagnie des Montres Longines, Francillon S.A. \t 0.733814001083374\n",
      "Rolex SA \t 0.7380605936050415\n",
      "Seiko Holdings Corporation \t 0.7396994233131409\n",
      "TAG Heuer S.A. \t 0.742731511592865\n",
      "Liverpool \t 0.7620262503623962\n",
      "Juventus \t 0.8283454775810242\n",
      "Real Madrid \t 0.8340783715248108\n",
      "Volvo \t 0.8571591973304749\n",
      "Android \t 0.9013322591781616\n",
      "OSX \t 0.9153567552566528\n",
      "Al Ahly \t 0.9355817437171936\n",
      "Windows \t 0.9544056057929993\n",
      "Linux Kernel \t 0.9668225646018982\n",
      "The Flu \t 1.16776442527771\n",
      "Headache \t 1.194204330444336\n",
      "Allergy \t 1.2166213989257812\n",
      "Conjunctivitis \t 1.226522445678711\n"
     ]
    }
   ],
   "source": [
    "similars, result = calculate_distances(test_embedding, docs_embedding_vectors, titles)\n",
    "for i in range(len(similars)):\n",
    "    print(result[i], '\\t',  similars[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8a47c2d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
